### DISCUSSION TITLE
Renaming classes and persistence

### DISCUSSION THREAD

**Discussion by mausch** [elsa-workflows/elsa-core#2773] [Q&A] [answered]
Hi, is there any documentation about what is safe to change / rename in the face of persisted Elsa entities?

For example, I see that the json blobs in my `workflow_instances` table contain things like:

```
      "Output": {
        "$id": "11",
        "$type": "MyNamespace.JobOrderModel, MyAssembly",
        "tenantId": "7b4831ad-9b28-48ca-b580-fb3a2be47ffa",
        "jobOrderId": "68e54645-f049-4d88-ad74-ae1400b059f9"
      },
```

So I'm guessing that if I rename my JobOrderModel class things will break?

Activities however only seem to be stored as their type name (without namespace or assembly), is that correct? So I'd be able to rename this in the code safely? How is this resolved from the type name alone though?

Or maybe there's already some facility in Elsa to do migrations and refactor things safely?

Many thanks

**Comment by sfmskywalker** [comment, marked as answer]
> So I'm guessing that if I rename my JobOrderModel class things will break?

That's right; Elsa relies on Newtonsoft.Json to (de)serialize models, and relies on the `$type` field to determine what type to instantiate.

> Activities however only seem to be stored as their type name (without namespace or assembly), is that correct? So I'd be able to rename this in the code safely? How is this resolved from the type name alone though?

For activities, it's a different story where a workflow definition's JSON format is deserialized into a `WorkflowDefinition`, which contains a list of `ActivityDefinition` objects, where each activity definition has a `Type: string` property, representing the type name of the activity (not the .NET type). This type name is the used by the `IWorkflowBlueprintMaterializer` to basically map a workflow definition into a workflow blueprint, using a lookup between activity type names and associated .NET type.

Regardless, if you rename an activity type, you will also need to update your workflow definition's JSON data to match the new name, **unless** you specified a custom `Type` value using the `ActivityAttribute` like so:

```csharp
[Activity(Type = "CustomNamedActivity")]
class MyActivity : Activity
{
}
```

In that case, you are free to rename your class, because the system will use `"CustomNamedActivity"` as the activity type name.
If you do not specify a custom type name explicitly, then the class name is used by default.
This also means that if you did specify a custom type name, changing it will again require you to update your workflow definitions.
